buffer/ring_growing.go

```diff
@@ -72,11 +86,44 @@ func (r *RingGrowing) WriteOne(data interface{}) {
 }
 
 // Len returns the number of items in the buffer.
-func (r *RingGrowing) Len() int {
+func (r *TypedRingGrowing[T]) Len() int {
 	return r.readable
 }
 
 // Cap returns the capacity of the buffer.
-func (r *RingGrowing) Cap() int {
+func (r *TypedRingGrowing[T]) Cap() int {
 	return r.n
 }
+
+// Ring is a dynamically-sized ring buffer which can grow and shrink as-needed.
+// Not thread safe.
+type Ring[T any] struct {
+	TypedRingGrowing[T]
```

Comment:
Embedding this type here is convenient to "inherit" its methods, but I'm not sure we want this in the public API. Or maybe that's actually a feature? If users have a `Ring` and want to a read such that the buffer definitely will not shrink, they could do `r.TypedRingGrowing.ReadOne()`.

Reply:
That particular usage seems a bit far-fetched and not like something that developers will discover if they need it - we would need to explicitly document the possibility.

I understand the convenience, but I have been bitten by wrappers embedding the type that they wrap when extending the type with new methods. Everything compiles fine even if the wrapper would have to be updated to also wrap the new methods.

It's cleaner to not embed the underlying type and instead repeat all methods as calls of the underlying method.

Reply:
Makes sense, done.